### ISSUE TITLE
[FR] A batch of FlagGems operators crash or produce wrong results on Ascend 910 NPU (pre-blacklisted by vllm-plugin-FL)

### ISSUE THREAD

**Issue by Lxiparer** [FlagOpen/FlagGems#5953]
## Summary

When running vLLM on Ascend 910 NPU through the `vllm-plugin-FL` dispatch layer, a set of FlagGems operators either **hard-crash the worker**, **hang the BiShengIR compilation pipeline**, or **produce numerically corrupted output**. These operators are currently worked around by an operator blacklist shipped inside `vllm-plugin-FL` (they fall back to the native `torch_npu` implementation). This issue documents them upstream so FlagGems can add Ascend-side guards / fixes rather than relying on a downstream blacklist.

## Environment

- FlagGems version: `5.3.4.post1.dev128+g05cd332b3`
- Backend toolchain: triton_ascend 3.2.1 / flagtree 0.6.0+ascend / BiShengIR
- CANN 9.0.0 (V100R001C10SPC001B250)
- torch 2.10.0 / torch_npu 2.10.0
- Hardware: Ascend 910 (product IT22HMDA_2_S), npu-smi 25.5.0
- Framework: vLLM 0.20.2 (V1 engine) + vllm-plugin-FL 0.2.0
- Model used to reproduce: Qwen3.6-35B-A3B (MoE + linear-attention hybrid), Phi-3-mini

### 1. Crashes (hard worker kill / segfault)

| Operator | Failure mode |
|----------|-------------|
| `lift_fresh` / `lift_fresh_copy` | Empty-tensor `coreDim=0`, grid=0 → Ascend runtime rejects → Fatal Python error, worker dies |
| `conv1d` / `conv2d` | Grouped/depthwise conv assertion failure on 2D varlen input |
| `topk` | Hard-crash NPU worker |
| `index` | Corrupts output (e.g. ViT rotary `cos[pos_ids]`) |
| `masked_scatter` / `masked_scatter_` / `scatter` / `scatter_` / `scatter_add_` / `gather` / `index_put` / `index_put_` / `_index_put_impl_` | Worker hard-crash / NPU segfault |
| `fill_scalar_` / `fill_tensor_` | ZeroDivisionError on zero-size slices |
| `argmax` | Invalid Runtime Function on NPU |

### 2. BiShengIR compilation failures / hangs

| Operator | Failure mode |
|----------|-------------|
| `empty_like` / `empty_strided` | BiShengIR compilation failure |
| `arange` | BiShengIR compilation timeout |
| `rand_like` | Hang in BiShengIR pipeline |
| `log` | BiShengIR compilation failure on some shapes |
| `mm` / `addmm` and elementwise arithmetic (`add`/`sub`/`mul`/`true_divide`/`floor_divide`/`reciprocal`/`neg`/`abs` ...) | Triton compilation hang on MoE 256-expert shapes |

### 3. Precision (wrong output, no crash)

| Operator | Failure mode |
|----------|-------------|
| `linear` | KernelGen Nvidia GEMM produces garbage on Ascend 910 → decode output is all `!!!!!` |
| `embedding` | Output corruption with autocast / mixed precision |

### `lift_fresh` — empty-tensor `coreDim=0`
- `src/flag_gems/ops/lift_fresh.py:76-77`
  ```python
  grid = lambda meta: (triton.cdiv(n_elements, meta["BLOCK_SIZE"]),)
  _copy_kernel[grid](x_contig, out, n_elements, BLOCK_SIZE=1024)
  ```
- For an empty tensor `numel()==0`, `grid` becomes `(0,)`. The Ascend runtime rejects `coreDim=0` and kills the process.
- Trigger: vLLM `attention.py` does `self.kv_cache = torch.tensor([])` under a `with target_device:` context → `torch.tensor` routes through `aten::lift_fresh`. torch 2.10 makes this path common.
- Suggested fix: early-return the input tensor (or a plain contiguous copy) when `n_elements == 0`, before launching the kernel.

### `conv1d`/`conv2d` — grouped conv assertion on 2D input
- `src/flag_gems/ops/conv2d.py:363-364`
  ```python
  input.shape[1] == groups * weight.shape[1]
  ), "Incompatible input ({input.shape}) and weights ({weight.shape}) shape with {groups} groups"
  ```
- FL's causal_conv1d path calls `F.conv1d(x, w.unsqueeze(1), ..., groups=dim)` with a **2D** `(dim, cu_seqlen)` input (depthwise, varlen). FlagGems routes conv1d into its conv2d, and the 2D input lacks the batch dim → assertion fails → request 500.
- Minor bug: the assertion message is a plain string, not an f-string (missing `f` prefix), so the actual shapes are never printed.
- Suggested fix: handle 2D depthwise/varlen conv1d input, or fall back to reference when `input.dim() < expected`.

### `linear` — KernelGen Nvidia GEMM on Ascend
- `src/flag_gems/ops/linear.py:15` — `# Generated by KernelGen`, `linear.py:38` `linear_kernel`, `linear.py:125` `def linear`.
- The Triton GEMM kernel was generated/tuned for Nvidia and is not validated on Ascend; it yields numerically garbage output (model decodes only `!`). Note the older `mm`/`addmm` blacklist did not cover `linear` because it is a separately registered op.
- Suggested fix: gate the KernelGen GEMM behind a validated-platform check, or provide an Ascend-correct path / reference fallback.

## Reproduction

1. Ascend 910 + FlagGems 5.3.4.post1.dev128, vLLM V1 engine.
2. Enable FlagGems for all ops (do NOT apply the FL blacklist).
3. Start a model that exercises these paths (Qwen3.6-35B-A3B for MoE/conv/linear; any model for `lift_fresh` via empty kv_cache init).
4. Observe: worker crash at startup (`lift_fresh` coreDim=0) or corrupted `!` output at decode (`linear`), depending on which op is hit first. Use `ASCEND_LAUNCH_BLOCKING=1` to get the true stack (async kernel errors otherwise mis-attribute the crash site).

## Current workaround

`vllm-plugin-FL` ships an operator blacklist (`vllm_fl/dispatch/config/ascend.yaml`, `flagos_blacklist`) that forces these ops back to native `torch_npu`. The stable config on the tested image contains 35 entries for this safety class (the original official 34 + `softmax`); the deployed container additionally pins `lift_fresh`/`lift_fresh_copy`/`conv2d` explicitly. This keeps the service stable and correct, but the underlying FlagGems kernels remain broken on Ascend. Filing upstream so the guards/fixes can live in FlagGems.
